apps/rush-lib/src/cli/utilities/pnpm/PnpmLinkManager.ts

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as fsx from 'fs-extra';
+import * as os from 'os';
+import * as path from 'path';
+import * as semver from 'semver';
+import encodeRegistry = require('encode-registry');
+
+import { JsonFile } from '@microsoft/node-core-library';
+
+import {
+  BaseLinkManager,
+  SymlinkKind
+} from '../base/BaseLinkManager';
+import Utilities from '../../../utilities/Utilities';
+import { BasePackage } from '../base/BasePackage';
+import { RushConstants } from '../../../RushConstants';
+import { IRushLinkJson } from '../../../data/RushConfiguration';
+import RushConfigurationProject from '../../../data/RushConfigurationProject';
+
+export class PnpmLinkManager extends BaseLinkManager {
+  protected _linkProjects(): Promise<void> {
+    return new Promise<void>((resolve: () => void, reject: (reason: Error) => void): void => {
+      // go ahead and find out the registry we are using, we need this to locate the pnpm store folder
+
+      // EXAMPLE: https://onedrive.pkgs.visualstudio.com/_packaging/odsp-npm/npm/registry
+      const registryUrl: string = Utilities.executeCommandAndCaptureOutput(
+        this._rushConfiguration.packageManagerToolFilename,
+        ['config', 'get', 'registry'],
+        this._rushConfiguration.commonTempFolder,
+        process.env);
+
+      // EXAMPLE: onedrive.pkgs.visualstudio.com
+      const encodedRegistry: string = encodeRegistry(registryUrl);
+
+      // EXAMPLE: c:/src/foo/common/temp/node_modules/.onedrive.pkgs.visualstudio.com
+      const registryPath: string = path.join(
+        this._rushConfiguration.commonTempFolder,
+        RushConstants.nodeModulesFolderName,
+        `.${encodedRegistry}`
+      );
+
+      try {
```

Comment:
>try [](start = 6, length = 3)

Shouldn't this cover the whole body?

Reply:
This entire function is synchronous; maybe you should just return Promise.resolve() at the end.

---
In reply to: [152449656](https://github.com/Microsoft/web-build-tools/pull/449#discussion_r152449656) [](ancestors = 152449656)